### PR TITLE
Adapt CAConstants for HIon

### PR DESCRIPTION
**Pull request by stahlleiton**
This PR include changes needed to avoid overflow while running the patatrack pixel tracking on HIon HLT menu using 2018 PbPb data.

The changes increase the thresholds of the following CA constants:

- maxNumberOfTuples : increased from 24 * 1024   to    32 * 1024 to avoid overflow on number of tuples.
- maxCellsPerHit : increased from 128   to    2 * 128 to avoid overflow on OuterHitOfCell.

#### PR validation:

The changes were tested on 2018 PbPb data using the HardProbe PDs.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

@fwyzard

### CODE REVIEW

File: RecoPixelVertexing/PixelTriplets/plugins/CAConstants.h [hunk 1]
```diff
@@ -32,9 +32,9 @@ namespace caConstants {
   constexpr uint32_t maxCellsPerHit = 128 / 2;
 #else   // GPU_SMALL_EVENTS
   // tested on MC events with 55-75 pileup events
-  constexpr uint32_t maxNumberOfTuples = 24 * 1024;
+  constexpr uint32_t maxNumberOfTuples = 32 * 1024;
```
Comment (fwyzard) [inline RecoPixelVertexing/PixelTriplets/plugins/CAConstants.h:35]:
```suggestion
  // and extended for Heavy Ions operations (24k -> 32k tuples, 128 -> 256 cells)
  constexpr uint32_t maxNumberOfTuples = 32 * 1024;
```

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
enable gpu

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
assign heterogeneous

**Comment by slava77** [comment]
@cmsbuild please test

it looks like the gpu tests never went through

**Comment by fwyzard** [comment]
From a simple test (building pixel triplets, GPU-only, pileup 50) this has a small but measurable impact.
```
2 CPUs:
  0: Intel(R) Xeon(R) Gold 6130 CPU @ 2.10GHz (16 cores, 32 threads)
  1: Intel(R) Xeon(R) Gold 6130 CPU @ 2.10GHz (16 cores, 32 threads)

1 visible NVIDIA GPUs:
  0: Tesla T4

Running 3 times over 10100 events with 2 jobs, each with 10 threads, 10 streams and 1 GPUs
```
  - reference: `1269.8 ±   3.6 ev/s`
  - with this PR: `1256.7 ±   4.0 ev/s`
  - impact: **`-1.0% ± 0.4%`**

**Comment by stahlleiton** [comment]
> From a simple test (building pixel triplets, GPU-only, pileup 50) this has a small but measurable impact.
> 
> ```
> 2 CPUs:
>   0: Intel(R) Xeon(R) Gold 6130 CPU @ 2.10GHz (16 cores, 32 threads)
>   1: Intel(R) Xeon(R) Gold 6130 CPU @ 2.10GHz (16 cores, 32 threads)
> 
> 1 visible NVIDIA GPUs:
>   0: Tesla T4
> 
> Running 3 times over 10100 events with 2 jobs, each with 10 threads, 10 streams and 1 GPUs
> ```
> 
>     * reference: `1269.8 ±   3.6 ev/s`
> 
>     * with this PR: `1256.7 ±   4.0 ev/s`
> 
>     * impact: **`-1.0% ± 0.4%`**

Thanks for the information.
Is it also possible to test the change of only one parameter at the time to see which of the two parameters impose the largest impact on the throughput?

**Comment by fwyzard** [comment]
The impact comes mostly from the change to the `maxCellsPerHit`:

constants | throughput
---|---
`maxNumberOfTuples = 24k`,  `maxCellsPerHit = 128` | `1269.8 ± 3.6 ev/s`
`maxNumberOfTuples = 24k`,  `maxCellsPerHit = 256` | `1258.9 ± 2.9 ev/s`
`maxNumberOfTuples = 32k`,  `maxCellsPerHit = 128` | `1265.2 ± 2.0 ev/s`
`maxNumberOfTuples = 32k`,  `maxCellsPerHit = 256` | `1256.7 ± 4.0 ev/s`

**Comment by stahlleiton** [comment]
Thanks @fwyzard .
Then one option to reduce the impact could be to try to template only maxCellsPerHit .

**Comment by fwyzard** [comment]
@VinInn, looking at [RecoPixelVertexing/PixelTriplets/plugins/gpuFishbone.h](https://github.com/cms-sw/cmssw/blob/master/RecoPixelVertexing/PixelTriplets/plugins/gpuFishbone.h), I'm wondering if the first part of the loop (the one that fills `x`, `y`, `z`, *etc.) can also be parallelised along `blockDim.x`, and those arrays moved into shared memory ?

If I followed the code correctly, even increasing `maxCellsPerHit` to `256` their combined size should be 22 bytes (4 floats, 1 int, 1 short) x 256 elements x 8 (blockDim.y) = 44k, which should fit in shared memory.

We may need to use cooperative groups to synchornise the sub-groups, but we can give it a try.

**Comment by slava77** [comment]
@fwyzard @stahlleiton 
please clarify on the status of this PR. From the last message about a week ago it sounded as if more developments are coming here.
Thank you.

**Comment by fwyzard** [comment]
hi @slava77, yes I think we want to explore a couple of other possibilities before making these changes.

**Comment by fwyzard** [comment]
Long term we should check if wecan speed up the code affected by `maxCellsPerHit`.
In the meantime could you check (or confirm) that we need to double it (from `128` to `256`), or if an intermediate value would be enough ?

**Comment by stahlleiton** [comment]
Sorry, seems I made a mistake while trying to go to a different release (the daily CMSSW development release I was using had already expired).

How do I go to a newer release without all commits appearing in the history?

**Pull request closed without merging**